### PR TITLE
Fixed singularity problem in Quaternion weightedMean

### PR DESCRIPTION
**Pull request by tomasmahlberg**
I have tested this in applied usage and it seems to get rid of the problems that were apparent before.

I could not find an issue for this TODO.

### CODE REVIEW

**Comment by tomasmahlberg** [comment]
@niklasborwell

**Comment by niklasborwell** [comment]
Looks good.

**Comment by marcusnaslund** [comment]
I don't understand how you use `preRotate` as a boolean statement. Isn't it always a quaternion, hence always not null, hence always true?

**Comment by marcusnaslund** [comment]
Wait, it's not a Quaternion. Ignore me. It's friday.

File: source/geometry/Quaternion.ooc [hunk 1]
```diff
@@ -234,7 +234,12 @@ Quaternion: cover {
 		// W - observationVectors
 		// Y - gibbsVector
 		// Z - vectorQuantityZ
-		// TODO: Fix singularity problem when the angle is close to PI, using sequential rotations
+		// Rotate quaternions if the angle is close to pi to avoid singularity problem
+		centerQuaternion := quaternions[quaternions count / 2]
+		preRotate := centerQuaternion angle(This identity) equals(Float pi, 0.5f)
+		if (preRotate)
+			for (index in 0 .. quaternions count)
+				quaternions[index] = centerQuaternion inverse * quaternions[index]
```
Comment (marcusnaslund) [inline source/geometry/Quaternion.ooc]: You calculate the inverse on every iteration now. Wouldn't it be better to do it just once?

**Comment by tomasmahlberg** [comment]
Updated with minor optimization.

**Pull request merged**